### PR TITLE
docs(dev-guide): developer guide

### LINKED ISSUE #45: Rules docs

### PR DESCRIPTION
**Pull request by buchslava**
Closes #45 #48

### CODE REVIEW

File: doc/developer-guide.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
+# ddf-validation Developer's guide 
+
+## Registry
```
Comment (rychkog) [inline doc/developer-guide.md]: What is registry (for me it is clear) but I think it should be explained in the doc.

File: doc/developer-guide.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+First of all new rule should be added to the [rules registry](../lib/ddf-rules/registry.js).
+
+A new rule should be added next way:
```
Comment (rychkog) [inline doc/developer-guide.md]: in the next way

File: doc/developer-guide.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+```javascript
+exports.MY_NEW_RULE = Symbol.for('MY_NEW_RULE');
+```
+
+Note: `Symbol.for` syntax is mandatory.
+
+Also, this rule should be added to descriptions:
```
Comment (rychkog) [inline doc/developer-guide.md]: Also description should be defined for a new rule as follows

File: doc/developer-guide.md (hunk 1 above)
Comment (buchslava) [inline doc/developer-guide.md]: Registry of issues types. Is it good?

File: doc/developer-guide.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+```javascript
+exports.descriptions = {
+  /// ...
+  [exports.MY_NEW_RULE]: 'Description for this rule',
+  /// ...
+};
+```
+
+## Rules applying explanation
```
Comment (rychkog) [inline doc/developer-guide.md]: Types of rules

File: doc/developer-guide.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+There are next types of rules:
```
Comment (rychkog) [inline doc/developer-guide.md]: There are 5 types of rules supported by ddf-validation

File: doc/developer-guide.md [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+* Index rules
+* Concept rules
+* Entity rules
+* Data points rules
+
+### General rules
+
+This kind of rules is about situation when program doesn't know about Concepts, Entities and Data points.
+In fact, this kind of rule is not DDF based. For example, rule `Folder is not DDF`.
+
```
Comment (buchslava) [inline doc/developer-guide.md]: This kind of rules describes cases that not depends on Concepts, Entities and Data points, ie depends on DDF generally.

File: doc/developer-guide.md [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+### Index rules
+
+This kind of rules is exclusively about `ddf--index` file. For example, `Index is not found`
+or `File in index doesn't exists`.
```
Comment (rychkog) [inline doc/developer-guide.md]: doesn't exist

File: doc/developer-guide.md (hunk 1 above)
Comment (rychkog) [inline doc/developer-guide.md]: Registry is a structure that contains all available rules and their description

File: doc/developer-guide.md (hunk 6 above)
Comment (rychkog) [inline doc/developer-guide.md]: General rules are applied to whole DDF data set. And the don't validate Concepts, Entities and Datapoints on their own.

File: doc/developer-guide.md [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+`Entity rules` are placed [here](../lib/ddf-rules/entity-rules.js).
+
+### Data points rules
+
+This kind of rules is exclusively about Data points. For example, `Unexpected time value`.
+
+`Data points` are placed [here](../lib/ddf-rules/data-point-rules.js).
+
+## Issue object
```
Comment (rychkog) [inline doc/developer-guide.md]: Issue is a class that describes single rule violation

File: doc/developer-guide.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+// ....
+const issue = new Issue(registry.MY_NEW_RULE);
+```
+
+Constructor `Issue` has only one argument - type of issue in accordance with the registry.
+
+Next section will be based on [this document](https://github.com/jsigbiz/spec) that needed for
+methods signatures describing.
+
+Also `Issue` class has methods for additional data filling:
```
Comment (rychkog) [inline doc/developer-guide.md]: With a help of following methods you can populate `Issue` instance with additional information regarding rule violation

File: doc/developer-guide.md [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+
+* Fill additional information (data) regarding this issue:
+
+```javascript
+fillData : (
+  data: Object
+) => Issue
+```
+
+* Fill suggestions for this issue:
```
Comment (rychkog) [inline doc/developer-guide.md]: what is suggestion?

File: doc/developer-guide.md [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+```
+
+Constructor `Issue` has only one argument - type of issue in accordance with the registry.
+
+Next section will be based on [this document](https://github.com/jsigbiz/spec) that needed for
+methods signatures describing.
+
+Also `Issue` class has methods for additional data filling:
+
+* Fill path when issue was raised and returns an updated object:
```
Comment (buchslava) [inline doc/developer-guide.md]: set path contains location of the file where issue was found

File: doc/developer-guide.md [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+  .fillSuggestions(['bar']);
+// also you can use next class members:
+console.log(issue.type);
+console.log(issue.path);
+console.log(issue.data);
+console.log(issue.suggestions);
+console.log(issue.isWarning);
+```
+
+By the way, all rules should return an `Issue` or array of `Issue` type.
```
Comment (rychkog) [inline doc/developer-guide.md]: an issue or array of issues

File: doc/developer-guide.md [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+
+## DdfData
+
+This class describes DDF data set generally. Code is placed [here](../lib/ddf-definition/ddf-data.js)
+
+DdfData object creation is:
```
Comment (rychkog) [inline doc/developer-guide.md]: How to create `DdfData` instance

File: doc/developer-guide.md (hunk 10 above)
Comment (buchslava) [inline doc/developer-guide.md]: suggestions => is an object that describes how to solve found issue

File: doc/developer-guide.md [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+) => Issue
+```
+
+* Set `warning` flag for particular issue:
+
+```javascript
+warning() => Issue
+```
+
+Typical way of issue creation:
```
Comment (rychkog) [inline doc/developer-guide.md]: How to create an `Issue` instance

File: doc/developer-guide.md [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+console.log(issue.data);
+console.log(issue.suggestions);
+console.log(issue.isWarning);
+```
+
+By the way, all rules should return an `Issue` or array of `Issue` type.
+
+## DdfData
+
+This class describes DDF data set generally. Code is placed [here](../lib/ddf-definition/ddf-data.js)
```
Comment (rychkog) [inline doc/developer-guide.md]: in general

File: doc/developer-guide.md [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+) => void
+```
+
+* Get object that allows you to Concepts control.
+
+```javascript
+getConcept : () => Concept
+```
+
+* Get object that allows you to Entities control.
```
Comment (rychkog) [inline doc/developer-guide.md]: Get object that represents Dataset concepts

File: doc/developer-guide.md [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+  
+  ddfData.dismiss();
+});
+```
+
+## There are two ways of rule implementation
+
+There are two kinds of rule implementation:
+
+* Data point rule
```
Comment (rychkog) [inline doc/developer-guide.md]: Datapoint specific rule

File: doc/developer-guide.md [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
+* each other rule
```
Comment (rychkog) [inline doc/developer-guide.md]: Non datapoint specific rule

File: doc/developer-guide.md [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
+
+    return result;
+  },
+  // ...
+};
+```
+
+### Data point rule creation
+
+This kind of rule has only one difference from other: second argument: `dataPointDetail`
```
Comment (rychkog) [inline doc/developer-guide.md]: from other -> comparing to other rules

File: doc/developer-guide.md [hunk 20]
```diff
@@ -0,0 +1,260 @@
...
+that contains all needed information about current set(file) data points.
```
Comment (rychkog) [inline doc/developer-guide.md]: of datapoints

**Pull request merged**